### DISCUSSION TITLE
How long it take for onSubAck callback to invoke incase of there is issue with the network or incase of any other issue

### DISCUSSION THREAD

**Discussion by niranjankk12** [aws/aws-iot-device-sdk-cpp-v2#661] [Q&A] [answered]
I am using below function to subscribe to a topic, we need to pass a callback function to get callback for server acknowledgement.

               /**
                 * Subscribes to topicFilter. OnMessageReceivedHandler will be invoked from an event-loop
                 * thread upon an incoming Publish message. OnSubAckHandler will be invoked
                 * upon receipt of a suback message.
                 *
                 * @param topicFilter topic filter to subscribe to
                 * @param qos maximum qos client is willing to receive matching messages on
                 * @param onMessage callback to invoke when a message is received based on matching this filter
                 * @param onSubAck callback to invoke with the server's response to the subscribe request
                 *
                 * @return packet id of the subscribe request, or 0 if the attempt failed synchronously
                 */
                uint16_t Subscribe(
                    const char *topicFilter,
                    QOS qos,
                    OnMessageReceivedHandler &&onMessage,
                    OnSubAckHandler &&onSubAck) noexcept;

**Question 1:**
The function declaration doesn't clarify how long one should wait to get callback, sometime it take more than 5 minutes (if internet connection is lost), we can't wait infinitely as there are chances that it can block the thread forever, 

**Question 2**   

Do we need to wait for call back if function return 0, there are chances of crash if local function address is passed to callback if the control is return from that function. Also when subscribe called on disconnected mqtt connection object, function return non-zero value but callback never get called.

**Comment by bretambrose** [comment, marked as answer]
1. The wait is unbounded unless you set an ack timeout when connecting: https://github.com/awslabs/aws-crt-cpp/blob/main/include/aws/crt/mqtt/MqttConnection.h#L257
2. 0 means failure and no callback will ever be invoked.

**Comment by niranjankk12** [comment]
I tired setting protocolOperationTimeoutMs to 10000 and tried calling Subscribe() multiple times when there is no internet connection, but as the return value of function (packet Id) is increases the time out also increases, when packet Id reaches to 20, function wait for more than 5 minutes. 

How the packet Id is related to a time out ?

**Comment by bretambrose** [comment]
The ack timeout is only for packets that have been sent and are waiting for a response.  If there's no connection, nothing will be sent, and there will be no timeout.
